### PR TITLE
runtime-rs: support watchable mount

### PR DESCRIPTION
**Pull request by liubin**
Use watchable mount to support inotify for virtio-fs.

Fixes: #5184 

Signed-off-by: Bin Liu <bin@hyper.sh>

### CODE REVIEW

File: src/libs/kata-types/src/k8s.rs [hunk 1]
```diff
@@ -12,18 +12,37 @@ use std::str::FromStr;
...
+const K8S_CONFIGMAP: &str = "kubernetes.io~configmap";
+// K8S_SECRET is the k8s specific path for `secret` volumes
+const K8S_SECRET: &str = "kubernetes.io~secret";
 
 /// Check whether the path is a K8S empty directory.
+pub fn is_empty_dir<P: AsRef<Path>>(path: P) -> bool {
+    is_special_dir(path, K8S_EMPTY_DIR)
+}
+
+/// Check whether the path is a K8S empty directory.
```
Comment (liubin) [inline src/libs/kata-types/src/k8s.rs:25]: configmap

File: src/libs/kata-types/src/k8s.rs [hunk 2]
```diff
@@ -12,18 +12,37 @@ use std::str::FromStr;
...
+pub fn is_configmap<P: AsRef<Path>>(path: P) -> bool {
+    is_special_dir(path, K8S_CONFIGMAP)
+}
+
+/// Check whether the path is a K8S empty directory.
```
Comment (liubin) [inline src/libs/kata-types/src/k8s.rs:30]: secret

File: src/agent/src/watcher.rs [hunk 3]
```diff
@@ -35,7 +35,9 @@ const MAX_SIZE_PER_WATCHABLE_MOUNT: u64 = 1024 * 1024;
 const WATCH_INTERVAL_SECS: u64 = 2;
 
 /// Destination path for tmpfs
-const WATCH_MOUNT_POINT_PATH: &str = "/run/kata-containers/shared/containers/watchable/";
+const WATCH_MOUNT_POINT_PATH: &str =
+    "/run/kata-containers/shared/containers/passthrough/watchable/";
+// FIXME: hardcode passthrough
```
Comment (liubin) [inline src/agent/src/watcher.rs:40]: Should consider 3 types:

- golang runtime path
- runtime-rs passthrough path
- runtime-rs rafs path(not available yet)

File: src/libs/kata-types/src/k8s.rs [hunk 4]
```diff
@@ -77,10 +96,119 @@ pub fn container_type_with_id(spec: &oci::Spec) -> (ContainerType, Option<String
...
+pub fn is_watchable_mount<P: AsRef<Path>>(path: P) -> bool {
+    if !is_secret(&path) && !is_configmap(&path) {
+        return false;
+    }
+
+    // we have a cap on number of FDs which can be present in mount
+    // to determine if watchable. A similar Check exists within the agent,
+    // which may or may not help handle case where extra files are added to
+    // a mount after the fact
+    let count = count_files(&path, 8).unwrap_or(0);
```
Comment (quanweiZhou) [inline src/libs/kata-types/src/k8s.rs:141]: Can you explain why the limit is set to 8, and maybe defining it as const is better.

Comment (liubin) [reply src/libs/kata-types/src/k8s.rs:141]: The original PR is https://github.com/kata-containers/kata-containers/pull/1964, the design is compatible with the golang runtime and agent. So here used 8 and the driver is `watchable-bind`.

File: src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs [hunk 5]
```diff
@@ -45,6 +56,60 @@ impl ShareFsMount for VirtiofsShareMount {
...
+                .join(PASSTHROUGH_FS_DIR)
+                .join("watchable")
+                .join(file_name)
+                .into_os_string()
+                .into_string()
+                .unwrap();
+
+            // kataWatchableBindDevType     = "watchable-bind"
+            let watchable_volume: Storage = Storage {
+                driver: String::from("watchable-bind"),
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs:89]: Yes, it's right, `watchable-bind`/`watchable` are valuable to be defined as a const, but `bind` is too common, so I'd like to keep the raw &str.

File: src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs [hunk 6]
```diff
@@ -45,6 +56,60 @@ impl ShareFsMount for VirtiofsShareMount {
...
+
+            fs::create_dir_all(&watchable_host_path).context(format!(
+                "unable to create watchable path: {:?}",
+                &watchable_host_path,
+            ))?;
+
+            fs::set_permissions(watchable_host_path, fs::Permissions::from_mode(0o750))?;
+
+            // path: /run/kata-containers/shared/containers/passthrough/watchable/config-map-name
+            let file_name = Path::new(&guest_path).file_name().unwrap();
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs]: fixed

File: src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs [hunk 7]
```diff
@@ -45,6 +56,60 @@ impl ShareFsMount for VirtiofsShareMount {
...
+
+        // watchable mounts
+        if is_watchable_mount(&config.source) {
+            // Create path in shared directory for creating watchable mount:
+            let host_rw_path = utils::get_host_rw_shared_path(&self.id);
+
+            // "/run/kata-containers/shared/sandboxes/$sid/rw/passthrough/watchable"
+            let watchable_host_path = Path::new(&host_rw_path)
+                .join(PASSTHROUGH_FS_DIR)
+                .join("watchable");
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs]: fixed

File: src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs [hunk 8]
```diff
@@ -45,6 +56,60 @@ impl ShareFsMount for VirtiofsShareMount {
...
+            let watchable_guest_mount = Path::new(KATA_GUEST_SHARE_DIR)
+                .join(PASSTHROUGH_FS_DIR)
+                .join("watchable")
+                .join(file_name)
+                .into_os_string()
+                .into_string()
+                .unwrap();
+
+            // kataWatchableBindDevType     = "watchable-bind"
+            let watchable_volume: Storage = Storage {
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/runtime-rs/crates/resource/src/share_fs/virtio_fs_share_mount.rs]: fixed.

**Review by quanweiZhou — APPROVED** [review]
Thanks, @liubin LGTM.

File: src/agent/src/watcher.rs [hunk 9]
```diff
@@ -34,9 +34,16 @@ const MAX_SIZE_PER_WATCHABLE_MOUNT: u64 = 1024 * 1024;
 /// How often to check for modified files.
 const WATCH_INTERVAL_SECS: u64 = 2;
 
-/// Destination path for tmpfs
+/// Destination path for tmpfs for golang runtime
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/agent/src/watcher.rs]: fixed, that path is used by the golang runtime

File: src/agent/src/watcher.rs [hunk 10]
```diff
@@ -34,9 +34,16 @@ const MAX_SIZE_PER_WATCHABLE_MOUNT: u64 = 1024 * 1024;
...
 const WATCH_MOUNT_POINT_PATH: &str = "/run/kata-containers/shared/containers/watchable/";
 
+/// Destination path for tmpfs for runtime-rs passthrough file sharing
+const WATCH_MOUNT_POINT_PATH_PASSTHROUGH: &str =
+    "/run/kata-containers/shared/containers/passthrough/watchable/";
+
+/// Destination path for tmpfs for runtime-rs rafs file sharing
+const WATCH_MOUNT_POINT_PATH_RAFS: &str = "/run/kata-containers/shared/containers/rafs/watchable/";
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/agent/src/watcher.rs]: fixed, deleted the rafs directory.

**Comment by liubin** [comment]
@Tim-0731-Hzt @studychao @bergwolf PTAL

**Review by Tim-0731-Hzt — APPROVED** [review]
Thanks @liubin  lgtm

**Pull request merged**